src/hotspot/share/code/codeBlob.cpp

```diff
@@ -73,82 +73,58 @@ unsigned int CodeBlob::allocation_size(CodeBuffer* cb, int header_size) {
-#ifdef COMPILER1
-  // probably wrong for tiered
-  assert(frame_size() >= -1, "must use frame size or -1 for runtime stubs");
-#endif // COMPILER1
-}
-#endif
-
-CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, int size, int header_size, int relocation_size,
-                   int content_offset, int code_offset, int frame_complete_offset, int data_offset,
-                   int frame_size, ImmutableOopMapSet* oop_maps, bool caller_must_gc_arguments) :
-  _oop_maps(oop_maps),
-  _name(name),
-  _size(size),
-  _header_size(header_size),
-  _relocation_size(relocation_size),
-  _content_offset(content_offset),
-  _code_offset(code_offset),
-  _frame_complete_offset(frame_complete_offset),
-  _data_offset(data_offset),
-  _frame_size(frame_size),
-  S390_ONLY(_ctable_offset(0) COMMA)
-  _kind(kind),
-  _caller_must_gc_arguments(caller_must_gc_arguments)
-{
-  DEBUG_ONLY( verify_parameters(); )
-}
-
-CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, int header_size,
-                   int frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments) :
+CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
+                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments) :
   _oop_maps(nullptr), // will be set by set_oop_maps() call
   _name(name),
   _size(size),
-  _header_size(header_size),
   _relocation_size(align_up(cb->total_relocation_size(), oopSize)),
-  _content_offset(CodeBlob::align_code_offset(_header_size + _relocation_size)),
+  _content_offset(CodeBlob::align_code_offset(header_size + _relocation_size)),
   _code_offset(_content_offset + cb->total_offset_of(cb->insts())),
-  _frame_complete_offset(frame_complete_offset),
   _data_offset(_content_offset + align_up(cb->total_content_size(), oopSize)),
   _frame_size(frame_size),
   S390_ONLY(_ctable_offset(0) COMMA)
+  _header_size(header_size),
+  _frame_complete_offset(frame_complete_offset),
   _kind(kind),
   _caller_must_gc_arguments(caller_must_gc_arguments)
 {
-  DEBUG_ONLY( verify_parameters(); )
+  assert(is_aligned(_size,            oopSize), "unaligned size");
+  assert(is_aligned(header_size,      oopSize), "unaligned size");
+  assert(is_aligned(_relocation_size, oopSize), "unaligned size");
+  assert(_data_offset <= _size, "codeBlob is too small: %d > %d", _data_offset, _size);
+  assert(code_end() == content_end(), "must be the same - see code_end()");
+#ifdef COMPILER1
+  // probably wrong for tiered
+  assert(_frame_size >= -1, "must use frame size or -1 for runtime stubs");
+#endif // COMPILER1
 
   set_oop_maps(oop_maps);
 }
 
 // Simple CodeBlob used for simple BufferBlob.
-CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, int size, int header_size) :
+CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, int size, uint16_t header_size) :
```

Comment:
Just a drive-by comment.  You might be able to use delegating constructors for CodeBlob so you don't have to have the field initializations twice.  Maybe the same for nmethod ?

Reply:
Thank you, @coleenp, foe looking on these changes.

Which fields are initialized twice? Only `_oop_maps` is set to `nullptr` before we proper build oop maps in first constructor.

The only saving could be lines of code but then I would have to check that `cb != nullptr` and do other additional checks which I don't think will save much lines.

Separation of `nmethod` constructor for native wrappers is helping clear see the difference and I would like to keep them separate. We have `init_defaults()` method for similar code and I can move more code into it from both constructors.

Reply:
Delegating constructors are the answer to having some common 'init' functions.  It would simply save lines of code that look the same in both constructor initializer lists.  But it's a drive-by comment.

Reply:
Okay.

Reply:
It is tempting to do for `nmethod` to replace `init_defaults()`. I will look what can be done.

Reply:
It does not work. It does not allow fields initialization after delegation:
```
src/hotspot/share/code/nmethod.cpp: In constructor 'nmethod::nmethod(Method*, CompilerType, int, int, CodeOffsets*, CodeBuffer*, int, ByteSize, ByteSize, OopMapSet*)':
src/hotspot/share/code/nmethod.cpp:1232:56: error: mem-initializer for 'nmethod::<unnamed union>::<unnamed struct>::_native_receiver_sp_offset' follows constructor delegation
```

Yes, I can move some fields initialization into body - but then it will be duplicated initialization.
From reading net, the delegated constructor should take the largest number of arguments which default values are set by delegating constructors. It is not easy applicable to `nmethod` constructors - it will add more lines of code than remove. `nmethod` constructors we have are too big and different enough that delegation will not improve code.

I will keep `init_defaults()` and extend it to reduce common code in constructors.